plugins/zenoh-plugin-storage-manager/src/replica/storage.rs

```diff
@@ -186,35 +261,235 @@ impl StorageService {
+        let mut tombstones = self.tombstones.write().await;
+        tombstones.insert(key_expr, timestamp);
+        if self.capability.persistence.eq(&Persistence::Durable) {
+            // flush to disk to makeit durable
+            let mut serialized_data = HashMap::new();
+            for (k, ts) in tombstones.key_value_pairs() {
+                serialized_data.insert(k, *ts);
+            }
+            if let Err(e) = std::fs::write(
+                zenoh_home().join(TOMBSTONE_FILENAME),
+                serde_json::to_string_pretty(&serialized_data).unwrap(),
+            ) {
+                error!("Saving tombstones failed: {}", e);
+            }
+        }
+        drop(tombstones);
+    }
 
+    async fn register_wildcard_update(&self, sample: Sample) {
+        // @TODO: change into a better store that does incremental writes
+        let key = sample.clone().key_expr;
+        let mut wildcards = self.wildcard_updates.write().await;
+        wildcards.insert(
+            &key,
+            Update {
+                kind: sample.kind,
+                data: StoredData {
+                    value: sample.value,
+                    timestamp: sample.timestamp.unwrap(),
+                },
+            },
+        );
+        if self.capability.persistence.eq(&Persistence::Durable) {
+            // flush to disk to makeit durable
+            let mut serialized_data = HashMap::new();
+            for (k, update) in wildcards.key_value_pairs() {
+                serialized_data.insert(k, serialize_update(update));
+            }
+            if let Err(e) = std::fs::write(
+                zenoh_home().join(WILDCARD_UPDATES_FILENAME),
+                serde_json::to_string_pretty(&serialized_data).unwrap(),
+            ) {
+                error!("Saving wildcard updates failed: {}", e);
+            }
+        }
+        drop(wildcards);
+    }
+
+    async fn is_deleted(&self, key_expr: &OwnedKeyExpr, timestamp: &Timestamp) -> bool {
+        // check tombstones to see if it is deleted in the future
+        let tombstones = self.tombstones.read().await;
+        let weight = tombstones.weight_at(key_expr);
+        weight.is_some() && weight.unwrap() > timestamp
+    }
+
+    async fn ovderriding_wild_update(
+        &self,
+        key_expr: &OwnedKeyExpr,
+        timestamp: &Timestamp,
+    ) -> Option<Update> {
+        // check wild card store for any futuristic update
+        let wildcards = self.wildcard_updates.read().await;
+        let mut ts = timestamp;
+        let mut update = None;
+        for node in wildcards.intersecting_keys(key_expr) {
+            let weight = wildcards.weight_at(&node);
+            if weight.is_some() && weight.unwrap().data.timestamp > *ts {
+                // if the key matches a wild card update, check whether it was saved in storage
+                // remember that wild card updates change only existing keys
+                let mut storage = self.storage.lock().await;
```

Comment:
Lock could be acquired below, just before `match storage.get(stripped_key, "")...`